tmt/lint.py

```diff
@@ -0,0 +1,308 @@
+LinterCallback = Callable[['Lintable'], LinterReturn]
+
+
+@dataclasses.dataclass(init=False)
+class Linter:
+    """ A single linter """
+
+    callback: LinterCallback
+    id: str
+
+    def __init__(self, callback: LinterCallback) -> None:
+        self.callback = callback
+
+        if not callback.__doc__:
+            raise tmt.utils.GeneralError(f"Linter '{callback}' lacks docstring.")
+
+        id, _ = textwrap.dedent(callback.__doc__).strip().split(':', 1)
+
+        self.id = id.strip()
+
+
+class Lintable(Generic[LintableT]):
+    """ Mixin class adding support for linting of class instances """
+
+    # Declare linter registry as a class variable, but do not initialize it. If initialized
+    # here, the mapping would be shared by all classes, which is not a desirable attribute.
+    # Instead, mapping will be created by `get_linter_registry()`.
+    _linter_registry: ClassVar[List[Linter]]
+
+    # Keep this method around, to correctly support Python's method resolution order.
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+
+    # Cannot use @property as this must remain classmethod
+    @classmethod
+    def get_linter_registry(cls) -> List[Linter]:
+        """ Return - or initialize - linter registry """
+
+        if not hasattr(cls, '_linter_registry'):
+            cls._linter_registry = []
+
+        return cls._linter_registry
+
+    @classmethod
+    def discover_linters(cls) -> None:
+        """
+        Discover and register all linters implemented by this class.
+
+        A linter is a method whose name starts with ``lint_`` prefix. It must
+        have a docstring which serves as a hint for ``--list-checks`` output.
+        """
+
+        for name in dir(cls):
+            if not name.startswith('lint_'):
+                continue
+
+            cls.get_linter_registry().append(tmt.lint.Linter(getattr(cls, name)))
+
+    @classmethod
+    def resolve_enabled_linters(
+            cls,
+            enable_checks: Optional[List[str]] = None,
+            disable_checks: Optional[List[str]] = None
+            ) -> List[Linter]:
+        linters: List[Linter] = []
+
+        if not enable_checks:
+            linters = cls.get_linter_registry()
+
+        else:
+            linters = []
+
+            for needle in enable_checks:
+                linters += [
+                    linter
+                    for linter in cls.get_linter_registry()
+                    if needle in linter.id
+                    ]
+
+        if disable_checks:
+            linters = [
+                linter for linter in linters
+                if linter.id not in disable_checks
+                ]
+
+        return linters
+
+    def lint(
+        self,
+        enable_checks: Optional[List[str]] = None,
+        disable_checks: Optional[List[str]] = None,
+        enforce_checks: Optional[List[str]] = None
+            ) -> Tuple[bool, List[LinterRuling]]:
```

Comment:
this indentation looks weird, I'm surprised flake did not complain, IMO would look better on one level or flipped - 8 spaces for the args and 4 for the return type

Reply:
Fixed in a9a62baf.